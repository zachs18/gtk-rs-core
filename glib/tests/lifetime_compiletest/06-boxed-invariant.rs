#![allow(unused)]

use std::sync::Once;

#[derive(Clone, Copy)]
#[repr(C)]
pub struct InnerRaw {
    data: *const [u8],
}

glib::wrapper! {
    pub struct TestBoxedRecord<'data>(Boxed<InnerRaw>);

    match fn {
        copy => |ptr| Box::into_raw(Box::new(*ptr)),
        free => |ptr| drop(Box::from_raw(ptr)),
    }
}

impl<'data> TestBoxedRecord<'data> {
    fn new(data: &'data [u8]) -> Self {
        unsafe { glib::translate::from_glib_none::<*const InnerRaw, Self>(&InnerRaw { data }) }
    }

    fn data(&self) -> &'data [u8] {
        unsafe { &*self.inner.data }
    }
}

fn not_covariant<'a>(record: TestBoxedRecord<'static>) -> TestBoxedRecord<'a> {
    record
}

fn main() {}
