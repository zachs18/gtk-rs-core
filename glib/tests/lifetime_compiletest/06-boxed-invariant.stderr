error: lifetime may not live long enough
  --> tests/lifetime_compiletest/06-boxed-invariant.rs:31:5
   |
30 | fn not_covariant<'a>(record: TestBoxedRecord<'static>) -> TestBoxedRecord<'a> {
   |                  -- lifetime `'a` defined here
31 |     record
   |     ^^^^^^ returning this value requires that `'a` must outlive `'static`
   |
   = note: requirement occurs because of the type `TestBoxedRecord<'_>`, which makes the generic argument `'_` invariant
   = note: the struct `TestBoxedRecord<'data>` is invariant over the parameter `'data`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
